sycl/source/device_selector.cpp

```diff
@@ -157,5 +161,152 @@ int host_selector::operator()(const device &dev) const {
   return Score;
 }
 
+namespace ONEAPI {
+namespace detail {
+struct filter {
+  backend Backend;
+  RT::PiDeviceType DeviceType;
+  int DeviceNum;
+  bool hasBackend = false;
+  bool hasDeviceType = false;
+  bool hasDeviceNum = false;
+  int MatchesSeen = 0;
+};
+
+std::vector<std::string> tokenize(std::string filter, std::string delim) {
+  std::vector<std::string> Tokens;
+  size_t Pos = 0;
+  std::string Input = filter;
+  std::string Tok;
+
+  while ((Pos = Input.find(delim)) != std::string::npos) {
+    Tok = Input.substr(0, Pos);
+    Input.erase(0, Pos + delim.length());
```

Comment:
Several allocations, deallocations and memcpy could be avoided here if we use something like StringView.

Reply:
Maybe - but it's a lot easier if I can chop up the string.

Perhaps someone smarter at C++ string magic can show a better way.

Reply:
@keryell, maybe you can suggest a good way to do it using C++14  ?

Reply:
Nowadays I use mostly C++20 ranges or range v3 for this kind of code similar to https://www.quora.com/How-do-I-parse-string-by-specific-delimiter